File: java/library/src/com/v2soft/styxlib/library/io/StyxDataReader.java
package com.v2soft.styxlib.library.io;

import java.io.UnsupportedEncodingException;

import com.v2soft.styxlib.library.types.ULong;

public class StyxDataReader implements IStyxDataReader {
    protected static final int sDataBufferSize = 16;
    protected byte [] mInternalBuffer;
    protected IStyxBuffer mBuffer;

    public StyxDataReader(IStyxBuffer buffer) {
        mInternalBuffer = new byte[getInternalBufferSize()];
        mBuffer = buffer;
    }    

    protected long readInteger(int bytes) {
        long result = getInteger(bytes);
        mBuffer.moveReadPointerBy(bytes);
        return result;
    }
    protected long getInteger(int bytes) {
        if ( bytes > getInternalBufferSize() ) throw new ArrayIndexOutOfBoundsException("Too much bytes to read");
        long result = 0L;
        int shift = 0;
        int readed = mBuffer.get(mInternalBuffer, 0, bytes);
        if ( readed != bytes ) throw new ArrayIndexOutOfBoundsException("Can't read bytes");
        for (int i=0; i<bytes; i++) {
            long b = (mInternalBuffer[i]&0xFF);
            if (shift > 0)
                b <<= shift;
            shift += 8;
            result |= b;
        }
        return result;
    }
    // ================================================================
    // IStyxDataReader methods
    // ================================================================
    @Override
    public String readUTFString() throws UnsupportedEncodingException {
        int count = readUInt16();
        byte[] bytes = new byte[count];
        read(bytes, 0, count);
        return new String(bytes, StyxDataWriter.sUTFCharset);
    }
    @Override
    public short readUInt8() {return (short) (readInteger(1)&0XFF);}
    @Override
    public int readUInt16() {return (int) (readInteger(2)&0xFFFF);}
    @Override
    public long readUInt32() {return (readInteger(4) &0xFFFFFFFF);}
    @Override
    public ULong readUInt64() {
        byte[] bytes = new byte[ULong.ULONG_LENGTH];
        read(bytes, 0, ULong.ULONG_LENGTH);
        return new ULong(bytes);        
    }
    @Override
    public int read(byte[] data, int offset, int count) {
        return mBuffer.read(data, offset, count);
    }
    @Override
    public long getUInt32() {return getInteger(4) & 0xFFFFFFFFL;}

    public int getInternalBufferSize() {
        return sDataBufferSize;
    }
}

File: native/V2StyxLib/lib/io/StyxDataReader.cpp
/*
 * StyxDataReader.cpp
 *
 *  Created on: May 29, 2012
 *      Author: mrco
 */

#include "io/StyxDataReader.h"

StyxDataReader::StyxDataReader(IStyxBuffer *buffer)
	: mBuffer(buffer){
}

StyxDataReader::~StyxDataReader() {
}

uint64_t StyxDataReader::readInteger(const size_t bytes) {
	uint64_t result = 0L;
	int shift = 0;
	const int q = 10;
	uint8_t mDataBuffer[q];
	read(mDataBuffer, 0, bytes);
	for (size_t i=0; i<bytes; i++) {
		long b = (mDataBuffer[i]&0xFF);
		if (shift > 0)
			b <<= shift;
		shift += 8;
		result |= b;
	}
	return result;
}

std::string StyxDataReader::readUTFString() {
	size_t count = readUInt16();
	char* bytes = new char[count];
	read((uint8_t*)bytes, 0, count);
	std::string result(bytes, count);
	delete [] bytes;
	return result;
}

uint8_t StyxDataReader::readUInt8() {
	return (short) (readInteger(1)&0XFF);
}

uint16_t StyxDataReader::readUInt16() {
	return (int) (readInteger(2)&0xFFFF);
}

uint32_t StyxDataReader::readUInt32() {
	return (readInteger(4) &0xFFFFFFFF);
}

uint64_t StyxDataReader::readUInt64() {
	return readInteger(8);
}

